Add env var option to allow multi-threading in onnx-mlir

Jiaqing-ASU opened the pull request:
Add one new option env var named 'menvVarName' in CompilerOptions and adjust the inputs of function 'ParseCommandLineOptions' in onnx-mlir.cpp. A python testing example named 'multi-threading-test.py' is provided under /docs/mnist_example/.

Signed-off-by: Jiaqing Chen <jchen501@asu.edu>

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented on docs/mnist_example/README.md:
```diff
@@ -106,6 +106,56 @@ onnx-mlir -O3 --EmitJNI mnist.onnx
+import os
+import threading
+ 
+def execCmd(cmd):
```
I guess that works here, aside of the security risk of letting the user choose any exec. I would hardwire onnx-mlir and/or check that the command start with onnx-mlir.

While that approach works, I was suggesting that the system commands occurs in `omCompileFromFile`.

In fact, I am advocating that we deprecate/remove all of the 
```C++
ONNX_MLIR_EXPORT int64_t omSetCompilerOptionsFromEnv(const char *envVarName);
ONNX_MLIR_EXPORT int64_t omSetCompilerOptionsFromArgs(
    int64_t argc, char *argv[]);
ONNX_MLIR_EXPORT int64_t omSetCompilerOptionsFromArgsAndEnv(
    int64_t argc, char *argv[], const char *envVarName);
ONNX_MLIR_EXPORT int64_t omSetCompilerOption(
    const OptionKind kind, const char *val);
ONNX_MLIR_EXPORT void omClearCompilerOption(const OptionKind kind);
ONNX_MLIR_EXPORT void omClearCompilerOption(const OptionKind kind);
```

The new c interface would then be: 
```C++
ONNX_MLIR_EXPORT int64_t omCompileFromFile(const char *inputFilename,
    const char *outputBaseName, EmissionTargetType emissionTarget,
    const char *flags,
    const char **outputFilename, const char **errorMessage);
```

and inside you can simply set a custom env variable (unique name) to the passed flags, and then launch "onnx-mlir --custom-env-var=<uniquename> <input file> -o <output file>"

The preferred way to to use something like this (this one is for llc, but you can do the same for onnx-mlir
```C++
  Command llvmToObj(/*exePath=*/!llcPath.empty() ? llcPath : kLlcPath);
  int rc = llvmToObj.appendStr(getOptimizationLevelOption())
               .appendStr(getTargetTripleOption())
               .appendStr(getTargetArchOption())
               .appendStr(getTargetCPUOption())
               .appendList(getXllcOption())
               .appendStr(getLLVMOption())
               .appendStr("-filetype=obj")
               .appendStr("-relocation-model=pic")
               .appendList({"-o", modelObjNameWithExt})
               .appendStr(bitcodeNameWithExt)
               .exec();
  return rc != 0 ? CompilerFailureInLLVMToObj : CompilerSuccess;
```

Once you have this function, you can change how the python works (and a few example in the code).

If you want help/me to do this infrastructure work,  I can do it.

Jiaqing-ASU replied:
I think I get it. Let me first have a try on this. Thank you.

gongsu832 commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
@jenkins-droid test this please

gongsu832 commented:
@jenkins-droid test this please

gongsu832 commented:
@jenkins-droid test this please

gongsu832 commented:
@jenkins-droid test this please

gongsu832 commented on src/onnx-mlir.cpp:
```diff
@@ -13,78 +13,29 @@
+      std::cerr << errorMessage << std::endl;
+    }
+    return rc;
   }
```
@Jiaqing-ASU I don't really understand what you are trying to do here. `onnx-mlir.cpp:main()` is the entry point of `onnx-mlir` executable. It's calling `omCompileFromFileViaCommand`, which from what I can see is calling `onnx-mlir` again. So you are asking `onnx-mlir` to call itself again and again.

This is why your PR is failing on all the Jenkins CI servers because they all run out of memory and Jenkins gets killed. On MacOS, it's failing because the system couldn't spawn more process anymore. Windows failure is different. It failed because function `strtok_r` doesn't exist on Windows (we need the Windows equivalent) but it would have failed similarly if `strtok_r` existed.

In any case, the modification to `onnx-mlir.cpp:main()` is completely wrong from what I can see. What you really need is to study a bit on `llvm::cl::opt` related stuff and see if there is any way to put all the options in thread local storage (i.e., each thread gets its own copy of the options).

Jiaqing-ASU commented on src/onnx-mlir.cpp:
```diff
@@ -58,10 +58,9 @@ int main(int argc, char *argv[]) {
   llvm::cl::ParseCommandLineOptions(argc, argv,
-      "ONNX-MLIR modular optimizer driver\n", nullptr,
-      OnnxMlirEnvOptionName.c_str());
+      "ONNX-MLIR modular optimizer driver\n", nullptr, customEnvFlags.c_str());
```
If inside the new C interface function named omCompileFromFileViaCommand, I give a name based on the input model file name before onnxmlirCompile.exec(), will this still be a potential problem? Actually from my understanding, since using the `onnx-mlir` exec as a target, it will work for multi-threading and multi-processing, if I use the onnxmlirCompile.exec() to run it, no matter what is the name comes from, right?

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented on docs/mnist_example/README.md:
```diff
@@ -106,6 +106,53 @@ onnx-mlir -O3 --EmitJNI mnist.onnx
 
+## Multi-threading
+
+onnx-mlir provides a multi-thread safe parallel compilation mode. Whether each thread is given a name or not by the user, onnx-mlir is multi-threaded safe. If you would like to give a name to a thread, use the `-customEnvFlags` keyword and an example can be found as follows.
```
> If you would like to give a name to a thread

I don't understand this. I though that argument was to replace the default `ONNX_MLIR_FLAGS` env var? In any case, because of how you constructed the `omCompileFromFileViaCommand` command, I am not sure we need to override the `ONNX_MLIR_FLAGS` env var anymore.

We can talk if easier to communicate on this.

Jiaqing-ASU replied:
My pleasure to work on this. Thanks for your nice support and understanding since I know I take a long time to get familiar with it and finish it. I think I have already fixed most of the comments.

For this one, I was thinking that many examples under [test](https://github.com/onnx/onnx-mlir/tree/35c1a375f7d57645d86a7d7bcab244213416f0a3/test) folder pass a custom name which repalce the original `OnnxMlirEnvOptionName` in `onnx-mlir.cpp` while calling the `llvm::cl::ParseCommandLineOptions` function. So I want to keep such an interface to replace the default `ONNX_MLIR_FLAGS` env var for future use. That's why I didn't directly remove the `-customEnvFlags`.

Also I am happy to discuss via Slack but I think I have trouble accessing the Slack channel via the link given in the onnx-mlir README and I am not able to find you. Could you please add me there or somewhere else? My email account of Slack is jchen501@asu.edu. Many thanks.

AlexandreEichenberger commented on src/CMakeLists.txt:
```diff
@@ -17,5 +17,4 @@ add_onnx_mlir_executable(onnx-mlir
   CompilerUtils
-  )
-  
+  )
```
please add the empty line again.

AlexandreEichenberger commented on docs/doc_example/main.cpp:
```diff
@@ -6,29 +6,24 @@
+    std::cerr << "Wrong program name received." << std::endl;
+    return -1;
+  }
+  for (int i = 1; i < argc; i++) {
```
Just to simplify the reading of the code, you could isolate the processing of the argc and args into a function that returns a single std::string. It would help the reading of the code.

you have it already below: `readArgsFromCommandLine`... just copy and paste

AlexandreEichenberger commented on docs/mnist_example/README.md:
```diff
@@ -106,6 +106,53 @@ onnx-mlir -O3 --EmitJNI mnist.onnx
+ 
+if __name__ == '__main__':
+    
+    # define 2 different commands
```
Eventually, we should hide this into the PyRuntime interface, but let's do it in a subsequent PR

AlexandreEichenberger commented on include/OnnxMlirCompiler.h:
```diff
@@ -110,6 +110,27 @@ ONNX_MLIR_EXPORT void omClearCompilerOption(const OptionKind kind);
+ *  @param emissionTarget Target format to compile to.
+ *  @param outputFilename Output file name of the compiled output for the given
+ * emission target. User is responsible for freeing the string.
+ *  @param flags A char * contains all the options provided to compile the
```
outstanding doc, maybe just add "space separated" list of options.

AlexandreEichenberger commented on src/onnx-mlir.cpp:
```diff
@@ -87,4 +86,4 @@ int main(int argc, char *argv[]) {
 
   return compileModule(module, context, outputBaseName, emissionTarget);
-}
+}
```
same here, git seems to like end of line after last statement

AlexandreEichenberger commented on src/Compiler/CompilerUtils.hpp:
```diff
@@ -68,4 +190,4 @@ int compileModule(mlir::OwningOpRef<mlir::ModuleOp> &module,
 std::string getTargetFilename(
     const std::string filenameNoExt, EmissionTargetType target);
-} // namespace onnx_mlir
+} // namespace onnx_mlir
```
add empty line please

AlexandreEichenberger commented on src/Compiler/ExternalUtil.hpp.in:
```diff
@@ -3,10 +3,15 @@
 
 namespace onnx_mlir {
-const std::string kExecPath = "@CMAKE_INSTALL_PREFIX@/bin/$<TARGET_FILE_NAME:onnx-mlir>"; /* fallback if not found by getExecPath */
+const std::string kExecPath =
```
very nice

AlexandreEichenberger commented on test/compilerlib/CompilerLibTest.cpp:
```diff
@@ -57,21 +59,37 @@ void readCommandLineAndKeepUnused(int &argc, char *argv[]) {
+// Read the arguments to pass to the testFileName and outputBaseName, other
+// arguments may then be processed by the ONNX-MLIR compiler.
+void readArgsFromCommandLine(int &argc, char *argv[]) {
+  if (std::string(argv[0]) != "onnx-mlir") {
```
I don't see why we would test the arg0 here. You can pass `argc` by value, no need for the `&`.

AlexandreEichenberger commented on test/compilerlib/CompilerLibTest.cpp:
```diff
@@ -90,4 +108,4 @@ int main(int argc, char *argv[]) {
   }
   return retVal;
-}
+}
```
end of line

AlexandreEichenberger reviewed (commented):
Outstanding work, really appreciate you working on this during your semester. Its almost there. 

Feel free to ping me on slack if you want to talk.

AlexandreEichenberger commented on docs/doc_example/main.cpp:
```diff
@@ -6,29 +6,26 @@
   const char *compiledFilename;
-  rc = onnx_mlir::omCompileFromFile("add.onnx", "add-cppinterface",
-      onnx_mlir::EmitLib, &compiledFilename, &errorMessage);
+  std::string commandLineString = readArgs(argc, argv);
```
awesome, much more readable, especially for documented examples

AlexandreEichenberger reviewed (approved):
LGTM, thanks for your dedication to get this one going through the multiple modifications. Let me know if you are ready for this one to land, or you still have some changes you want to push through.

Jiaqing-ASU commented:
> LGTM, thanks for your dedication to get this one going through the multiple modifications. Let me know if you are ready for this one to land, or you still have some changes you want to push through.

I do not have other stuff to push but I do see some checks were not successful and I am looking into it. Maybe I need some help with that since I do not think my most recent changes will lead to those errors. But before I did my most changes, all the stuff ran successfully under the Docker environment. I have no idea about the error:

FAILED: docs/doc_example/add.so 
cd /workdir/onnx-mlir/build/docs/doc_example && /usr/bin/python3.8 /workdir/onnx-mlir/build/docs/doc_example/gen_add_onnx.py && /workdir/onnx-mlir/build/Debug/bin/onnx-mlir -O3 add.onnx

error: Target architecture is unknown: No available targets are compatible with triple "@ONNX_MLIR_DEFAULT_TRIPLE@"
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
Segmentation fault

Could you please let me know how should I fix this issue?

AlexandreEichenberger commented:
@gongsu832 do you mind to look at the problem this PR is having? It fails to compile because of this
```
error: Target architecture is unknown: No available targets are compatible with triple "@ONNX_MLIR_DEFAULT_TRIPLE@"
```

I looked at changes done to CMakeList but did not find any obvious source of issues.

@Jiaqing-ASU if we find no obvious issue, the only think I can recommend is to update to the latest to both your main and this branch, copy the changed files somewhere else, and the make a new branch out of main, make sure it compile, and then put back some of your changes bit by bits.

Jiaqing-ASU commented:
> @Jiaqing-ASU if we find no obvious issue, the only think I can recommend is to update to the latest to both your main and this branch, copy the changed files somewhere else, and the make a new branch out of main, make sure it compile, and then put back some of your changes bit by bits.

@AlexandreEichenberger Thanks so much. This is exactly what I was thinking. I think I will do this if we are not familiar with this error message.

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
updated to the latest

Jiaqing-ASU commented:
This ran successfully in my environment after I made some changes. But I just ran rebuild `cmake` under the `llvm-project/build` and `onnx-mlir/build` each time. I did check the `build` folder which changed each time after I rebuilt. Maybe I do need to first remove the whole `build` folder and then rebuild everything. Let me redo this and find the place that leads to this error. Thanks.

AlexandreEichenberger commented:
Will give it a try.

Jiaqing-ASU commented:
> FAILED: docs/doc_example/add.so cd /workdir/onnx-mlir/build/docs/doc_example && /usr/bin/python3.8 /workdir/onnx-mlir/build/docs/doc_example/gen_add_onnx.py && /workdir/onnx-mlir/build/Debug/bin/onnx-mlir -O3 add.onnx
> 
> error: Target architecture is unknown: No available targets are compatible with triple "@ONNX_MLIR_DEFAULT_TRIPLE@" qemu: uncaught target signal 11 (Segmentation fault) - core dumped Segmentation fault

This error has been solved. It was caused by the `static` functions.

AlexandreEichenberger commented:
@Jiaqing-ASU hope you don't mind, I looked at your code and fixed a few minor issues, mainly compile header file (for default tool location) and a few tweaks to launch the onnx-mlir process (adding the input model and the requested output file name). Removed one redundant function (getTool) that I was able to reuse from elsewhere. Kept your code as close as you wrote as possible.

Please have a look and if you are happy, we will put it down. Thanks for the great work.

Jiaqing-ASU commented:
> @Jiaqing-ASU hope you don't mind, I looked at your code and fixed a few minor issues, mainly compile header file (for default tool location) and a few tweaks to launch the onnx-mlir process (adding the input model and the requested output file name). Removed one redundant function (getTool) that I was able to reuse from elsewhere. Kept your code as close as you wrote as possible.
> 
> 
> 
> Please have a look and if you are happy, we will put it down. Thanks for the great work.

@AlexandreEichenberger It looks good to me. Thanks so much for the help. I really appreciate it. After this PR is merged, I will modify the Python interface accordingly and submit a new PR hopefully this week or maybe early next week. Many thanks!

sorenlassen commented on src/Compiler/CompilerUtils.cpp:
```diff
@@ -152,90 +153,82 @@ static std::string getToolPath(std::string tool) {
-        _args({llvm::sys::path::filename(_path).str()}) {}
+// Append a single optional string argument.
+Command &Command::appendStrOpt(const llvm::Optional<std::string> &arg) {
+  if (arg.hasValue())
+    _args.emplace_back(arg.getValue());
```
@Jiaqing-ASU can you please change `hasValue` and `getValue` to `has_value` and `value` instead? that will silence a deprecation compiler warning (which we've been working to weed out across the code base)

sorenlassen replied:
see [this recent fix](https://github.com/onnx/onnx-mlir/pull/1622/files#diff-34127a063216e23b47bcb21b2b8dfe936b159241c0774480d4501593d5224798R173-R174)

AlexandreEichenberger commented on src/Compiler/OnnxMlirCompiler.cpp:
```diff
@@ -49,6 +53,50 @@ ONNX_MLIR_EXPORT const char *omGetCompilerOption(const OptionKind kind) {
   return strdup(val.c_str());
 }
 
+#ifdef _WIN32
```
@sstamenova I know its already pushed, if you have a preferred way to handle MS custom handling, please let me know and I will integrate in a follow up cleanup PR.

sstamenova replied:
My recommendation would be for any situation like this to first look for the equivalent function in llvm - they almost always exist and they are properly implemented cross-platform, so we wouldn't need any windows specific handling.

If that can't be avoided, something like what you have here generally works, but in this particular case, I think there is an llvm function.
